extensions/data-transfer/portability-data-transfer-generic/src/main/java/org/datatransferproject/datatransfer/generic/GenericTransferExtension.java

```diff
@@ -0,0 +1,214 @@
+  }
+}
+
+class GenericTransferServiceConfig {
+  private final String serviceId;
+  private final URL endpoint;
+  private final List<GenericTransferServiceVerticalConfig> verticals;
+
+  public GenericTransferServiceConfig(
+      @JsonProperty(value = "serviceId", required = true) String serviceId,
+      @JsonProperty(value = "endpoint", required = true) URL endpoint,
+      @JsonProperty(value = "verticals", required = true)
+          List<GenericTransferServiceVerticalConfig> verticals) {
+    this.serviceId = serviceId;
+    this.endpoint = endpoint;
+    this.verticals = verticals;
+  }
+
+  public String getServiceId() {
+    return serviceId;
+  }
+
+  public URL getEndpoint() {
+    return endpoint;
+  }
+
+  public List<GenericTransferServiceVerticalConfig> getVerticals() {
+    return verticals;
+  }
+
+  public boolean supportsVertical(DataVertical vertical) {
+    for (GenericTransferServiceVerticalConfig verticalConfig : verticals) {
+      if (verticalConfig.getVertical().equals(vertical)) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+
+public class GenericTransferExtension implements TransferExtension {
+  Map<DataVertical, Importer<?, ?>> importerMap = new HashMap<>();
+
+  @Override
+  public boolean supportsService(String service) {
+    try {
+      TransferServiceConfig config = TransferServiceConfig.getForService(service);
+      if (config.getServiceConfig().isEmpty()) {
+        return false;
+      }
+      // Parse failures throw
+      parseConfig(config.getServiceConfig().get());
+    } catch (IOException e) {
+      return false;
+    }
+    // Found and parsed a valid generic service config for the service
+    return true;
+  }
+
+  @Override
+  public void initialize(ExtensionContext context) {
+    JobStore jobStore = context.getService(JobStore.class);
+    TransferServiceConfig configuration = context.getService(TransferServiceConfig.class);
+    if (configuration.getServiceConfig().isEmpty()) {
+      throw new RuntimeException("Empty service configuration");
+    }
+    GenericTransferServiceConfig serviceConfig;
+    try {
+      serviceConfig = parseConfig(configuration.getServiceConfig().get());
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException("Invalid service configuration", e);
+    }
+
+    AppCredentialStore appCredentialStore = context.getService(AppCredentialStore.class);
+    String serviceNameUpper = serviceConfig.getServiceId().toUpperCase();
+    AppCredentials appCredentials;
+    try {
+      appCredentials =
+          appCredentialStore.getAppCredentials(
+              format("%s_KEY", serviceNameUpper), format("%s_SECRET", serviceNameUpper));
+    } catch (IOException e) {
+      throw new RuntimeException(
+          format(
+              "Failed to get application credentials for %s (%s)",
+              serviceNameUpper, serviceConfig.getServiceId()),
+          e);
+    }
+
+    if (serviceConfig.supportsVertical(BLOBS)) {
```

Comment:
For better readability, it's worth extracting these conditions to their private methods. 

Even better,  if we can auto-update the `importerMap` these by iterating over the `DataVertical` entries. We may need to maintain a map of  DataVertical -> corresponding Serializer.

Reply:
I'm not sure I understand what you mean by extracting to private methods - could you add an example?

For auto-updating the importer map, we'd need a map of DataVertical -> GenericImporter, because sometimes we need GenericFileImporter and sometimes just GenericImporter (although they could be merged).
I gave it a go with BLOBS and CALENDAR in https://github.com/dtinit/data-transfer-project/commit/19db8804f9910bbe855b90525cbe3a621b8211a0 and I don't believe it's a significant improvement in readability, plus it's going to get messier when trying to tackle the PHOTOS/VIDEOS vs MEDIA logic, which is straightforward with the longer approach

Reply:
The length of this method increased and will continue to increase as we add more verticals. So, extracting to private methods that mutates the importer map would reduce the length of this method and makes it more readable. 

We can fix this later as well. This shouldn’t be a blocker for merging this PR.